### PR TITLE
Introduce a new example using the Tensorflow NumPy API to implement keras layers & models

### PR DESCRIPTION
**Pull request by LukeWood**

### CODE REVIEW

**Comment by LukeWood** [comment]
Let me know if `guides` is the wrong directory for this!

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR!

- I recommend making this a code example, under `examples/keras_recipes`.
- Before the PR is approved, please only include the `.py` in the PR. The autogenerated files are added as the very last step. So the PR should just contain `examples/keras_recipes/tf_numpy_models.py` at first.
- It seems you have included a large image file as part of the code, in base64. This prevents previewing the file on GitHub. Instead, please use an external host (like imgur) for any image you want to embed.

**Comment by LukeWood** [comment]
> I recommend making this a code example, under examples/keras_recipes.
Done!
> Before the PR is approved, please only include the .py in the PR. The autogenerated files are added as the very last step. So the PR should just contain examples/keras_recipes/tf_numpy_models.py at first.
Should I also include the jupyter notebook I used to generate the file?  If no I can remove it!

> It seems you have included a large image file as part of the code, in base64. This prevents previewing the file on GitHub. Instead, please use an external host (like imgur) for any image you want to embed.
Done!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+"""
+
+"""
+# Implementing a Keras Layer Based Model with TNP
+If desired, `tnp` can also be used in layer oriented Keras code structure.  Let's
+implement the same model, but using a layered approach!
+"""
+
+
+class DenseTNP(keras.layers.Layer):
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:146]: Other models here use `TNP` first, so here `TNPDense` to be consistent

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    def __init__(self, units):
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:147]: The API differs from `Dense`, in particular because this layer applies `relu` by default.

I would recommend making the activation configurable:

```python
if self.activation:
    return self.activation(outputs)
return outputs
```

And then define:

```python
def relu(inputs):
    return tnp.maximum(inputs, 0.)
```

And pass `relu` as the `activation` in `TNPDense`

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            shape=(self.units,),
+            initializer="random_normal",
+            trainable=True,
+        )
+
+    def call(self, inputs):
+        return tnp.maximum(tnp.matmul(inputs, self.w) + self.bias, 0)
+
+
+class LinearProjector(keras.layers.Layer):
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:169]: Once you make `activation` modular, this is just a default dense layer with 1 unit.

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    model.add(DenseTNP(3))
+    model.add(DenseTNP(3))
+    model.add(LinearProjector())
+    return model
+
+
+model = create_layered_tnp_model()
+model.compile(
+    optimizer="adam",
+    loss="mean_squared_error",
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:199]: Random though: maybe it would be useful to show how to define the loss in NumPy too

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+The Tensorboard also allows you to explore the computation graph used in your models.
+"""
+
+"""
+![Tensorboard graph exploration](https://i.imgur.com/tOrezDL.png)
+"""
+
+"""
+  The ability to introspect into your models can be valuable during debugging.
+"""
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:329]: You can use a single text block for everything above.

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+To test our models we will use the Boston housing prices regression dataset.
+"""
+
+(x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data(
+    path="boston_housing.npz", test_split=0.2, seed=113
+)
+
+
+def evaluate_model(model: keras.Model):
+    [loss, meanAbsolutePercentError] = model.evaluate(x_test, y_test, verbose=0)
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:56]: Always use snake case for variable names, e.g. `percent_error`

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the update! Looking great. We'll do a round of copyedits at a later stage.

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 6 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py:56]: Done!

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 4 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py:197]: Went ahead and added a section on this above 👍

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+    def call(self, inputs):
+        return tnp.matmul(inputs, self.w) + self.bias
+
+
+from keras.models import Sequential
+
+
+def create_layered_tnp_model():
+    model = Sequential()
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done!

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 2 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Great idea, this is much nicer!  Done

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 1 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 6 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 5 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py:317]: Done!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+from keras.models import Sequential
+from keras.layers import Dense
+
+
+def create_mixed_model():
+    model = Sequential()
+    model.add(DenseTNP(3))
+    # The model will have no issue using a normal Dense layer
+    model.add(Dense(3, use_bias=True))
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Ah I see - that explains the small discrepency in accuracies.  thanks!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
+"""
+Title: Writing Keras Models With Tensorflow NumPy
+Author: lukewood
+Date created: 2021/08/28
+Last modified: 2021/08/28
+Description: Overview of how to use the Tensorflow NumPy API to write Keras models.
+"""
+"""
+# Introduction
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:9]: Switch to `##` (level 2) for all section titles, only the example title (autogenerated) should be top-level

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    loss="mean_squared_error",
+    metrics=[keras.metrics.MeanAbsolutePercentageError()],
+)
+model.build((None, 13,))
+model.summary()
+
+evaluate_model(model)
+
+"""
+You can also seamlessly between `tnp` layers and native Keras layers!
```
Comment (fchollet) [inline examples/keras_recipes/tensorflow_numpy_models.py:216]: Missing word

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+"""
+
+"""
+# Setup
+Tensorflow NumPy requires TensorFlow 2.5 or later.
+"""
+
+import tensorflow as tf
+
+print(tf.__version__)
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py:32]: Ah yeah:

```
This call enables type promotion in TensorFlow and also changes type inference, when converting literals to tensors, to more strictly follow the NumPy standard.
```

I'll update this

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 3 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+import tensorflow as tf
+import tensorflow.experimental.numpy as tnp
+import keras
+import numpy as np
+
+"""
+Additionally, in order to use Tensorflow NumPy you must modify some global settings using
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Ah no it is not.  It is only for type promotion.  Thanks!

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+Title: Writing Keras Models With Tensorflow NumPy
+Author: lukewood
+Date created: 2021/08/28
+Last modified: 2021/08/28
+Description: Overview of how to use the Tensorflow NumPy API to write Keras models.
+"""
+"""
+# Introduction
+[NumPy](https://numpy.org/) is a hugely successful python linear algebra library.
+Many machine learning practitioners prototype projects NumPy, converting to Jax or
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: ah yeah - good point.  I updated this to say: 
```
Thanks to `tf_numpy`, you can write Keras layers or models in the NumPy style!
```

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+PyTorch at later stages.
+
+Tensorflow recently launched [tf_numpy](https://www.tensorflow.org/guide/tf_numpy), a
+TensorFlow copy of a subset of the NumPy API.  Thanks to `tf_numpy`, porting layers or
+models from NumPy to Keras is now easy!
+
+The Tensorflow NumPy API has full integration with the Tensorflow ecosystem.  Features
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: done, thanks!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        )
+
+    def call(self, inputs):
+        outputs = tnp.matmul(inputs, self.w) + self.bias
+        if self.activation:
+            return self.activation(outputs)
+        return outputs
+
+
+from keras.models import Sequential
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: done, thanks!

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 9 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Updated these to all be ##

File: examples/keras_recipes/tensorflow_numpy_models.py (hunk 10 above)
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Ah good catch - added

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+model.summary()
+
+evaluate_model(model)
+
+"""
+You can also seamlessly between `tnp` layers and native Keras layers!
+"""
+
+from keras.models import Sequential
+from keras.layers import Dense
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Ah thanks - fixed

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+
+First let's define a simple `TNPForwardFeedRegressionNetwork` class.
+"""
+
+
+class TNPForwardFeedRegressionNetwork(keras.models.Model):
+    def __init__(self, blocks=None):
+        super(TNPForwardFeedRegressionNetwork, self).__init__()
+        if not isinstance(blocks, list):
+            raise ValueError(f"blocks must be a list, got={blocks}")
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: updated - thanks!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+Description: Overview of how to use the Tensorflow NumPy API to write Keras models.
+"""
+"""
+# Introduction
+[NumPy](https://numpy.org/) is a hugely successful python linear algebra library.
+Many machine learning practitioners prototype projects NumPy, converting to Jax or
+PyTorch at later stages.
+
+Tensorflow recently launched [tf_numpy](https://www.tensorflow.org/guide/tf_numpy), a
+TensorFlow copy of a subset of the NumPy API.  Thanks to `tf_numpy`, porting layers or
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done - much better worded!  Thanks!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+
+In this example, we will implement a `Model` subclass that performs regression over the
+boston housing dataset using the `tnp` API.  Note that differentiation and gradient
+descent is handled automatically when using the `tnp` API alongside keras.
+
+First let's define a simple `TNPForwardFeedRegressionNetwork` class.
+"""
+
+
+class TNPForwardFeedRegressionNetwork(keras.models.Model):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done - updated!

File: examples/keras_recipes/tensorflow_numpy_models.py [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+    def __init__(self, blocks=None):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply examples/keras_recipes/tensorflow_numpy_models.py]: Done - good point!

**Review by fchollet — APPROVED** [review]
Excellent! Please add the generated files (command `python autogen.py add_example keras_recipes/tensorflow_numpy_models`). I pushed some minor formatting fixes, please pull them first.

**Comment by LukeWood** [comment]
@googlebot I consent.

**Comment by LukeWood** [comment]
Ah wait I think you need to comment that you consent @fchollet

Let me know if there is anything else that I need to do here!

**Comment by LukeWood** [comment]
ah wait a minute... I rebased incorrectly 🤦

**Comment by LukeWood** [comment]
> ah wait a minute... I rebased incorrectly 🤦

I was able to fix this by running `git rebase -i HEAD~7` and deleting the two offending commits out of my local history.  now it should be good!
